README.md
# Domoticz scripts
Collection of domoticz lua and shell scripts.

### Lua:
- `script_time_radar.lua`: Radar detection and alerting on specified roads (Netherlands only)
- `script_time_traffic.lua`: Traffic jam detection and alerting on specified routes

### Shell
Collection of domoticz lua scripts, currently consisting of:

- `message.sh`: Send a text message using a Telegram bot
- `snapshot.sh`: Send a photo using a Telegram bot

## Installation:
Put the scripts you want inside your domoticz scripts folders:

`domoticz/scripts/lua`
`domoticz/scripts/shell`

Keep the name of the lua script filenames the same, and they will be 
triggered every minute (actual code execution every x minutes can 
be set inside every script).

CHANGELOG.md
# Change Log - domoticz scripts
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).
Changelog based on guidelines from [Keep a CHANGELOG](http://keepachangelog.com/).

## 2.0.0
#### Changed:
- BREAKING: Moved lua scripts to lua folder, added shell folder.
#### Added:
- Shell scripts to send text and photos using Telegram bot.

## 1.0.2
#### Fixed:
- script_time_radar.lua had a bug, not sending messages.

## 1.0.1
#### Fixed:
- script_time_radar.lua script now uses alert virtual hardware device with status updates.
- script_time_traffic.lua script now uses alert virtual hardware device with status updates.

## 1.0.0
#### Added:
- script_time_radar.lua script.
- script_time_traffic.lua script.

shell/message.sh
#!/bin/sh

# Shell script to send a text message using Telegram
#
# See https://core.telegram.org/bots#creating-a-new-bot how to create a
# Telegram bot. Fill in your token in the TOKEN variable below.
#
# Then create a group in which you add the bot, and anyone you want to
# receive the messages. Then fetch the chat_id of your group (can be tricky,
# Google "telegram get chat_id"). Then enter the CHAT_ID below.
#
# The script accepts 1 parameter, which should be the text to be sent. If your
# text contains spaces, wrap it in quotes:
# messages.sh "this is a test"
#
# (c) 2017 Johnny Mijnhout

############## EDIT THESE SETTINGS TO YOUR PERSONAL VALUES #####################

TOKEN=<your:token>
CHAT_ID=<your_chat_id>

############################# END SETTINGS #####################################

TEXT=$1
curl -s "https://api.telegram.org/bot$TOKEN/sendMessage?chat_id=$CHAT_ID&text=$TEXT" > /dev/null &


shell/snapshot.sh
#!/bin/sh

# Shell script to send photos using Telegram
#
# See https://core.telegram.org/bots#creating-a-new-bot how to create a
# Telegram bot. Fill in your token in the TOKEN variable below.
#
# Then create a group in which you add the bot, and anyone you want to
# receive the messages. Then fetch the chat_id of your group (can be tricky,
# Google "telegram get chat_id"). Then enter the CHAT_ID below.
#
# (c) 2017 Johnny Mijnhout

############## EDIT THESE SETTINGS TO YOUR PERSONAL VALUES #####################

TOKEN=<your:token>
CHAT_ID=<your_chat_id>

############################# END SETTINGS #####################################

ROOT_DIR='/home/pi/Pictures/webcams'
DATE=`date +%Y%m%d`
TIME=`date +%H%M%S`
HOUR=`date +%H`
TEXT=$1
START_HOUR=$2
END_HOUR=$3

if [ ! -z "$START_HOUR" ] && [ ! -z "$END_HOUR" ]; then
	if [ "$START_HOUR" -lt "$END_HOUR" ]; then
		if [ "$HOUR" -lt "$START_HOUR" ] || [ "$HOUR" -gt "$END_HOUR" ]; then
			exit 0
		fi
	else
		if [ "$HOUR" -lt "$START_HOUR" ] && [ "$HOUR" -gt "$END_HOUR" ]; then
			exit 0
		fi
	fi
fi

if [ -z "$TEXT" ]; then
        TEXT="Activiteit op $DATE om $TIME"
fi

curl -s "https://api.telegram.org/bot$TOKEN/sendMessage?chat_id=$CHAT_ID&text=$TEXT" > /dev/null &

umask 000
mkdir -p $ROOT_DIR/$DATE

(avconv -rtsp_transport tcp -i 'rtsp://192.168.0.112:554/user=admin&password=&channel=1&stream=0.sdp?real_stream--rtp-caching=100' -f image2 -vframes 1 -pix_fmt yuvj420p $ROOT_DIR/$DATE/$TIME.garage.jpg > /dev/null 2>&1 &&
curl -s -X POST "https://api.telegram.org/bot$TOKEN/sendPhoto" -F chat_id="$CHAT_ID" -F caption="$TEXT" -F photo="@$ROOT_DIR/$DATE/$TIME.garage.jpg" > /dev/null) &

(avconv -rtsp_transport tcp -i 'rtsp://192.168.0.113:554/user=admin&password=&channel=1&stream=0.sdp?real_stream--rtp-caching=100' -f image2 -vframes 1 -pix_fmt yuvj420p $ROOT_DIR/$DATE/$TIME.voortuin.jpg > /dev/null 2>&1 &&
curl -s -X POST "https://api.telegram.org/bot$TOKEN/sendPhoto" -F chat_id="$CHAT_ID" -F caption="$TEXT" -F photo="@$ROOT_DIR/$DATE/$TIME.voortuin.jpg" > /dev/null) &

exit 0


lua/script_time_radar.lua
-- Domoticz script to check radars
--
-- This script uses the ANWB feed to check for radars in a list of specified
-- roads. If one or more radars are stationed on one of the roads, it will
-- send out a notification.
--
-- In Domotics create new hardware of type Dummy if you don't already
-- have it. After that, create a new virtual sensor, type = 'waarschuwing'. In
-- 'apparaten' you will find this new device. The number in the 'Idx'
-- column is your virtual_device_index.
--
-- Create a new user variable in 'gebruikers variablen'. Give it a name
-- and set type to 'integer'. The name you used is your user_variable_name.
--
-- (c) 2017 Johnny Mijnhout

------------ EDIT THESE SETTINGS TO YOUR PERSONAL VALUES ----------------------------

-- road names to check, enclosed and separated by pipe symbols(for example |A1|A201|)
roads="<-- your road list -->"

-- virtual text sensor index number
virtual_device_index = "<-- your index number -->"

-- user variable to store if a notification has been sent
user_variable_name = "<-- your user variable name -->"

-- check interval
check_internal_minutes = 15

-- hours between which you want to check
home_hour_start = 7
home_hour_end   = 9
work_hour_start = 15
work_hour_end   = 17

------------------------ END SETTINGS -----------------------------------------------

-- init commandArray
commandArray = {}

-- check if a message has been sent
message_sent = tonumber(uservariables[user_variable_name])

-- only run this script every x minutes on week days
day_number  = tonumber(os.date("%w"))
time_object = os.date("*t")
if time_object.min % check_internal_minutes > 0 or day_number == 0 or day_number == 6 then
  goto done
end

-- only check between speicified times
if (time_object.hour < home_hour_start or time_object.hour > home_hour_end) and (time_object.hour < work_hour_start or time_object.hour > work_hour_end) then
  if message_sent == 1 then
    commandArray["Variable:"..user_variable_name] = "0"
  end
  goto done
end

-- import json lib
json = (loadfile "/home/pi/domoticz/scripts/lua/JSON.lua")()

-- set defaults
message      = ""
total_radars = 0

-- get data from ANWB
json_data   = assert(io.popen("curl 'https://www.anwb.nl/feeds/gethf'"))
json_table  = json_data:read("*all")
json_data:close()
anwb = json:decode(json_table)
for k, v in pairs(anwb.roadEntries) do
  road = tostring(v.road)
  if string.match(roads,"|"..road.."|") then
    radar_number = # v.events.radars
    if radar_number > 0 then
      message = message..tostring(radar_number).." flitser(s) op de "..road.."! "
      total_radars = total_radars + radar_number
    end
  end
end

-- if no radars are found, set this in the message
if total_radars == 0 then
  message = "Geen flitsers gevonden op de "..string.sub(string.gsub(roads, "|", ", "), 3, -3)
end

-- set alert status
alert_status = total_radars + 1
if alert_status > 4 then
  alert_status = 4
end

-- append time to message
message = message.." (update "..tostring(os.date("%H:%M"))..")"

-- send message if radars are found
if total_radars > 0 and message_sent == 0 then
    commandArray["SendNotification"] = message
    commandArray["Variable:"..user_variable_name] = "1"
end

-- update virtual device text
commandArray["UpdateDevice"] = virtual_device_index.."|"..alert_status.."|"..tostring(message)

::done::
return commandArray

lua/script_time_traffic.lua
-- Domoticz script to check traffic jams
--
-- This script uses Google Directions API to check the duration of travel
-- between 2 geolocations at certain times. You can set the maximum travel
-- time in minutes, if the duration is longer than this time a notification
-- will be sent.
--
-- Get a Google Directions API key here:
-- https://developers.google.com/maps/documentation/directions/
-- and click "get a key".
--
-- Fetch your geolocations for example by searching for a location on
-- https://maps.google.com
-- You will find the geolocation after the @ symbol in the url.
--
-- In Domotics create new hardware of type Dummy if you don't already
-- have it. After that, create a new virtual sensor, type = 'waarschuwing'. In
-- 'apparaten' you will find this new device. The number in the 'Idx'
-- column is your virtual_device_index.
--
-- Create a new user variable in 'gebruikers variablen'. Give it a name
-- and set type to 'integer'. The name you used is your user_variable_name.
--
-- (c) 2017 Johnny Mijnhout

------------ EDIT THESE SETTINGS TO YOUR PERSONAL VALUES ----------------------------

-- location coordinates (for example 12.3456,65.4321)
home_coordinates = "<-- your home geo location -->"
work_coordinates = "<-- your work geo location -->"

-- Google Directions API key
google_api_key = "<-- your Google Directions API key -->"

-- virtual text sensor index number
virtual_device_index = "<-- your index number -->"

-- user variable to store if a notification has been sent
user_variable_name = "<-- your user variable name -->"

-- check interval
check_internal_minutes = 15

-- hours between which you want to check
home_hour_start = 7
home_hour_end   = 9
work_hour_start = 15
work_hour_end   = 17

-- maximum travel time
max_home_work_minutes = 60
max_work_home_minutes = 60

-- location names
home_name = "huis"
work_name = "werk"

------------------------ END SETTINGS -----------------------------------------------

-- init commandArray
commandArray = {}

-- check if a message has been sent
message_sent = tonumber(uservariables[user_variable_name])

-- only run this script every x minutes on week days
day_number  = tonumber(os.date("%w"))
time_object = os.date("*t")
if time_object.min % check_internal_minutes > 0 or day_number == 0 or day_number == 6 then
  goto done
end

-- decide time of day for direction
if time_object.hour >= home_hour_start and time_object.hour <= home_hour_end then
  origin_coordinates      = home_coordinates
  destination_coordinates = work_coordinates
  destination_name        = work_name
  max_travel_minutes      = max_home_work_minutes
elseif time_object.hour >= work_hour_start and time_object.hour <= work_hour_end then
  origin_coordinates      = work_coordinates
  destination_coordinates = home_coordinates
  destination_name        = home_name
  max_travel_minutes      = max_work_home_minutes
else
  -- reset message sent to receive new messages when travelling back
  if message_sent == 1 then
    commandArray["Variable:"..user_variable_name] = "0"
  end
  goto done
end

-- import json lib
json = (loadfile "/home/pi/domoticz/scripts/lua/JSON.lua")()

-- get traffic data from Google Directions API
json_data  = assert(io.popen("curl 'https://maps.googleapis.com/maps/api/directions/json?origin="..origin_coordinates.."&destination="..destination_coordinates.."&departure_time=now&key="..google_api_key.."'"))
json_table = json_data:read("*all")
json_data:close()
gmaps      = json:decode(json_table)

-- Read from the data table
distance_km      = math.ceil(gmaps.routes[1].legs[1].distance.value/1000)
duration_minutes = math.ceil(gmaps.routes[1].legs[1].duration_in_traffic.value/60)
summary_text     = gmaps.routes[1].summary

-- calculate delay
delay_minutes    = duration_minutes - max_travel_minutes
if delay_minutes < 0 then
  delay_minutes = 0
end

-- translate summary text
summary_text = string.gsub(summary_text, "and", "en")

-- create string with travel information
travel_text = tostring(duration_minutes).." min reistijd naar "..tostring(destination_name).." ("..tostring(distance_km).." km), via "..tostring(summary_text)

-- set message to be pushed and stored
if delay_minutes > 0 then
  message = "FILE! "..travel_text
  -- if no message has been sent, send out the notification
  if message_sent == 0 then
    commandArray["SendNotification"] = message
    commandArray["Variable:"..user_variable_name] = "1"
  end
  -- set alert status
  if delay_minutes > 30 then
      alert_status = 4
  elseif delay_minutes > 15 then
      alert_status = 3
  elseif delay_minutes > 0 then
      alert_status = 2
  end
else
  message = "Weg is vrij, "..travel_text
  -- if a message has been sent, send a new one that traffic has cleared
  if message_sent == 1 then
    commandArray["SendNotification"] = message
    commandArray["Variable:"..user_variable_name] = "0"
  end
  --reset alert status
  alert_status = 1
end
-- append time to message
message = message.." (update "..tostring(os.date("%H:%M"))..")"

-- update virtual device alert + counter
commandArray['UpdateDevice'] = virtual_device_index.."|"..alert_status.."|"..tostring(message)

::done::
return commandArray
